ddtrace/internal/datadog/profiling/dd_wrapper/src/uploader_builder.cpp

```diff
@@ -0,0 +1,141 @@
+#include "uploader_builder.hpp"
+#include "libdatadog_helpers.hpp"
+#include "types.hpp"
+
+#include <algorithm>
+#include <mutex>
+#include <numeric>
+#include <string_view>
+#include <utility>
+#include <vector>
+
+using namespace Datadog;
+
+void
+UploaderBuilder::set_env(std::string_view _env)
+{
+    if (!_env.empty())
+        env = _env;
+}
+void
+UploaderBuilder::set_service(std::string_view _service)
+{
+    if (!_service.empty())
+        service = _service;
+}
+void
+UploaderBuilder::set_version(std::string_view _version)
+{
+    if (!_version.empty())
+        version = _version;
+}
+void
+UploaderBuilder::set_runtime(std::string_view _runtime)
+{
+    if (!_runtime.empty())
+        runtime = _runtime;
+}
+void
+UploaderBuilder::set_runtime_version(std::string_view _runtime_version)
+{
+    if (!_runtime_version.empty())
+        runtime_version = _runtime_version;
+}
+void
+UploaderBuilder::set_profiler_version(std::string_view _profiler_version)
+{
+    if (!_profiler_version.empty())
+        profiler_version = _profiler_version;
+}
+void
+UploaderBuilder::set_url(std::string_view _url)
+{
+    if (!_url.empty())
+        url = _url;
+}
+void
+UploaderBuilder::set_tag(std::string_view _key, std::string_view _val)
+{
+
+    if (!_key.empty() && !_val.empty()) {
+        std::lock_guard<std::mutex> lock(tag_mutex);
+        user_tags[_key] = _val;
+    }
+}
+
+void
+UploaderBuilder::set_runtime_id(std::string_view _runtime_id)
+{
+    if (!_runtime_id.empty())
+        runtime_id = _runtime_id;
+}
+
+std::string
+join(const std::vector<std::string>& vec, const std::string& delim)
+{
+    return std::accumulate(
+      vec.begin(), vec.end(), std::string(), [&delim](const std::string& a, const std::string& b) -> std::string {
+          if (!a.empty()) // Use !a.empty() for clarity
```

Comment:
What is `b.empty()`?

Reply:
Ah, good catch.  empty `b`, nonempty `a` will result in spurious delimiters being appended.